dbms/src/Flash/Mpp/GRPCReceiveQueue.h

```diff
@@ -0,0 +1,235 @@
+    bool cancelWith(const String & reason)
+    {
+        auto ret = recv_queue.cancelWith(reason);
+        if (ret)
+            handleTheRemainingTags();
+        return ret;
+    }
+
+    bool finish()
+    {
+        auto ret = recv_queue.finish();
+        if (ret)
+            handleTheRemainingTags();
+        return ret;
+    }
+
+    // Pop data from queue and kick the CompletionQueue to re-push the data
+    // if there is data failing to be pushed before because the queue was full.
+    template <typename U>
+    bool pop(U && data)
+    {
+        auto ret = recv_queue.pop(std::forward<U>(data)) == MPMCQueueResult::OK;
+        if (ret)
+            kickCompletionQueue();
+        return ret;
+    }
+
+    // Push data into the queue.
+    //
+    // Return FULL if the queue is full and `new_tag` is saved.
+    // When the next pop/finish is called, the `new_tag` will be pushed
+    // into grpc completion queue.
+    template <typename U>
+    GRPCReceiveQueueRes push(U && data, void * new_tag)
+    {
+        auto res = recv_queue.tryPush(data);
+        switch (res)
+        {
+        case MPMCQueueResult::OK:
+            return GRPCReceiveQueueRes::OK;
+        case MPMCQueueResult::FINISHED:
+            return GRPCReceiveQueueRes::FINISHED;
+        case MPMCQueueResult::CANCELLED:
+            return GRPCReceiveQueueRes::CANCELLED;
+        case MPMCQueueResult::FULL:
+            // Handle this case later.
+            break;
+        default:
+            RUNTIME_ASSERT(false, log, "Result {} is invalid", static_cast<Int32>(res));
+        }
+
+        std::lock_guard lock(mu);
+
+        // Double check if this queue is full.
+        res = recv_queue.tryPush(data);
+        switch (res)
+        {
+        case MPMCQueueResult::OK:
+            return GRPCReceiveQueueRes::OK;
+        case MPMCQueueResult::FINISHED:
+            return GRPCReceiveQueueRes::FINISHED;
+        case MPMCQueueResult::CANCELLED:
+            return GRPCReceiveQueueRes::CANCELLED;
+        case MPMCQueueResult::FULL:
+        {
+            tags.push_back(new_tag);
+            return GRPCReceiveQueueRes::FULL;
+        }
+        default:
+            RUNTIME_ASSERT(false, log, "Result {} is invalid", magic_enum::enum_name(res));
+        }
+    }
+
+private:
+    friend class tests::TestGRPCReceiveQueue;
+
+    /// Wake up its completion queue.
+    void kickCompletionQueue()
+    {
+        {
+            std::lock_guard lock(mu);
+            if (tags.empty())
+                return;
+
+            transferTagToKickReceiveTagWithNoLock();
+        }
+
+        callKickFunc();
+    }
+
+    void handleTheRemainingTags()
+    {
+        std::lock_guard lock(mu);
+        while (!tags.empty())
+        {
+            transferTagToKickReceiveTagWithNoLock();
+            callKickFunc();
+        }
+    }
+
+    void transferTagToKickReceiveTagWithNoLock()
+    {
+        kick_recv_tag.pushTag(tags.front());
```

Comment:
Why not just put the tag to `kick_recv_tag` directly in `push()`?

Reply:
> Why not just put the tag to `kick_recv_tag` directly in `push()`?

Because we should ensure that each tag can be pushed into completion queue only once.